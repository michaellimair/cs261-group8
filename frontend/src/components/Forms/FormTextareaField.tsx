import {
  FormControl, FormHelperText, FormLabel, Textarea,
} from '@chakra-ui/react';
import { IFieldProps } from 'customTypes/form';
import { PropsWithChildren } from 'react';
import { useTranslation } from 'react-i18next';

interface IFormTextAreaFieldProps<T> extends Omit<IFieldProps<T>, 'type'> {
  label: string;
}

const FormTextAreaField = <T extends any>({
  name,
  autoComplete,
  register,
  error,
  required,
  label: formLabel,
  placeholder,
  helperText,
  disabled,
}: PropsWithChildren<IFormTextAreaFieldProps<T>>) => {
  const { t } = useTranslation();

  return (
    <FormControl
      id={name}
      data-testid={`${name}-formcontrol`}
      isRequired={required}
      isInvalid={Boolean(error)}
      mb={2}
    >
      <FormLabel data-testid={`${name}-label`}>{formLabel}</FormLabel>
      <Textarea
        disabled={disabled}
        placeholder={placeholder}
        autoComplete={autoComplete}
        {...register(name, {
          required: t('no_blank', { field: t('content') }) as any,
        })}
      />
      {helperText && (
      <FormHelperText>
        {helperText}
      </FormHelperText>
      )}
    </FormControl>
  );
};

export default FormTextAreaField;
