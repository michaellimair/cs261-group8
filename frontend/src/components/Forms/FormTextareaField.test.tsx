import { render, RenderResult } from '@testing-library/react';
import FormTextareaField from './FormTextareaField';

interface ISampleForm {
  field: string;
}

describe('components/FormTextareaField', () => {
  const fieldName: keyof ISampleForm = 'field';
  const autoComplete = 'email';
  const label = 'customlabel';
  const placeholder = 'testplaceholder';
  let result: RenderResult<typeof import('@testing-library/dom/types/queries'), HTMLElement>;
  let registerFn: jest.Mock;

  beforeEach(() => {
    registerFn = jest.fn();
    result = render(
      <FormTextareaField<ISampleForm>
        name={fieldName}
        autoComplete={autoComplete}
        register={registerFn}
        label={label}
        placeholder={placeholder}
      />,
    );
  });

  afterEach(() => {
    registerFn.mockReset();
    result.unmount();
  });

  it('renders correctly with the right set of attributes', async () => {
    const textareaFormControl = result.queryByTestId(`${fieldName}-formcontrol`)!;
    expect(textareaFormControl).not.toBeNull();
    expect(textareaFormControl).toMatchSnapshot();

    const inputLabelField = textareaFormControl.querySelector('label')!;
    expect(inputLabelField).toBeTruthy();
    expect(inputLabelField.innerHTML).toEqual(label);

    const selectField = textareaFormControl.querySelector('textarea')!;
    expect(selectField).toBeTruthy();
    expect(selectField.id).toEqual(fieldName);
    expect(selectField.getAttribute('autoComplete')).toEqual(autoComplete);
    expect(selectField).toMatchSnapshot();
  });
});
