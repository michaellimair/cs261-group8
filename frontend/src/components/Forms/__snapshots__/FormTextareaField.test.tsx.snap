// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FormTextareaField renders correctly with the right set of attributes 1`] = `
<div
  class="chakra-form-control css-wlea3r"
  data-testid="field-formcontrol"
  role="group"
>
  <label
    class="chakra-form__label css-uvho5g"
    data-testid="field-label"
    for="field"
    id="field-label"
  >
    customlabel
  </label>
  <textarea
    autocomplete="email"
    class="chakra-textarea css-0"
    id="field"
    placeholder="testplaceholder"
  />
</div>
`;

exports[`components/FormTextareaField renders correctly with the right set of attributes 2`] = `
<textarea
  autocomplete="email"
  class="chakra-textarea css-0"
  id="field"
  placeholder="testplaceholder"
/>
`;
